### PR TITLE
Added Breadth First Search in Typescript

### LINKED ISSUE #462: Breadth First Search in Typescript
This is an open issue for the Breadth-First Search algorithm implementation in TypeScript. If you wish to contribute, just make some noise and we will assign you.

### PR DESCRIPTION
**Pull request by SahanAmarsha**
**Fixes  #462**

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [x] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [x] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [x] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [x] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [x] This pull request shall only be merged once the build passes.

-   [x] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/gr-bfsrh/typescript/breadth_first_search.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        const visitedNodes: Array<number> = [];
+        let nodesToProcess: Array<number> = [0];
+
+        while (nodesToProcess.length) {
+            const node = nodesToProcess.shift();
+
+            // process the node if there is a callback
+            if (cb) {
+                const result = cb(node, visitedNodes, nodesToProcess);
+                if (result === true) break;
```
Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:30]: Codacy found an issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

File: algorithms/gr-bfsrh/typescript/breadth_first_search.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    graph.addEdge(0, 1);
+    graph.addEdge(0, 2);
+    graph.addEdge(1, 2);
+    graph.addEdge(2, 0);
+    graph.addEdge(2, 3);
+    graph.addEdge(3, 3);
+
+    console.log(JSON.stringify(graph.breadthFirstSearch()));
+
+}
```
Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:61]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

File: algorithms/gr-bfsrh/typescript/breadth_first_search.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        this.adjacencies = [];
+        this.adjacencies[0] = [];
+    }
+
+    addNode(node: number) {
+        this.adjacencies[node] = [];
+    }
+
+    addEdge(node1: number, node2: number) {
+        this.adjacencies[node1].push(node2)
```
Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:16]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

File: algorithms/gr-bfsrh/typescript/breadth_first_search.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    graph.addNode(3);
+
+    graph.addEdge(0, 1);
+    graph.addEdge(0, 2);
+    graph.addEdge(1, 2);
+    graph.addEdge(2, 0);
+    graph.addEdge(2, 3);
+    graph.addEdge(3, 3);
+
+    console.log(JSON.stringify(graph.breadthFirstSearch()));
```
Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:59]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

File: algorithms/gr-bfsrh/typescript/breadth_first_search.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    graph.addEdge(0, 1);
+    graph.addEdge(0, 2);
+    graph.addEdge(1, 2);
+    graph.addEdge(2, 0);
+    graph.addEdge(2, 3);
+    graph.addEdge(3, 3);
+
+    const result = JSON.stringify(graph.breadthFirstSearch());
+}
```
Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:62]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

File: algorithms/gr-bfsrh/typescript/breadth_first_search.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    graph.addEdge(0, 1);
+    graph.addEdge(0, 2);
+    graph.addEdge(1, 2);
+    graph.addEdge(2, 0);
+    graph.addEdge(2, 3);
+    graph.addEdge(3, 3);
+
+    var result = graph.breadthFirstSearch();
+}
```
Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:62]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

Comment (SahanAmarsha) [reply algorithms/gr-bfsrh/typescript/breadth_first_search.ts:62]: I believe this is a false error

Comment (RavinduSachintha) [inline algorithms/gr-bfsrh/typescript/breadth_first_search.ts:62]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6340523)

Comment (Danushka96) [reply algorithms/gr-bfsrh/typescript/breadth_first_search.ts:62]: @SahanAmarsha Add a semicolon at line 62,

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@SahanAmarsha can you add this contribution to the algorithm chart on readme file? don't change anything else.

**Comment by SahanAmarsha** [comment]
Done :v:

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@all-contributors please add @SahanAmarsha for code and doc

**Comment by RavinduSachintha** [comment]
@SahanAmarsha Thank you for your contribution. keep it up :+1: 
Make sure to star this repository :wink:

**Pull request merged**